Integrated server support for CORS

#### Perceived Problem

I am not able to use the `nexus-future` GraphQL server from my frontend. I am getting the error
```
Access to XMLHttpRequest at 'http://localhost:4000/graphql' from origin 'http://localhost:8100' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

#### Ideas / Proposed Solution(s)

- Allow users to enable CORS on the server, maybe through something like 

```ts
settings.change{{
  server: {
    cors: true;
  }
});
```

I added the following line to my `tsconfig.json`:

```
 "esModuleInterop": true
```

When I run `npx nexus build`, I get this new error:

```
npx nexus build
1141 ● nexus:plugin:nexus-plugin-prisma Running Prisma generators ...
2248 ● nexus:build Running typegen & extracting types from addToContext calls
 496 ● nexus:build Compiling a production build
Error: api/app.ts:13:18 - error TS7016: Could not find a declaration file for module 'cors'. '/Users/crabasa/Code/nexus-test/node_modules/cors/lib/index.js' implicitly has an 'any' type.
  Try `npm install @types/cors` if it exists or add a new declaration (.d.ts) file containing `declare module 'cors';`

13 import cors from 'cors'
                    ~~~~~~

    at Object.compile (/Users/crabasa/Code/nexus-test/node_modules/nexus/src/lib/tsc.ts:162:11)
    at Object.buildNexusApp (/Users/crabasa/Code/nexus-test/node_modules/nexus/src/lib/build/build.ts:99:3)
    at Build.parse (/Users/crabasa/Code/nexus-test/node_modules/nexus/src/cli/commands/build.ts:32:5)
    at main (/Users/crabasa/Code/nexus-test/node_modules/nexus/src/cli/main.ts:86:18)
```

I'm sorry if I'm missing something, but I'm new to TypeScript and it seems to be what Nexus is built with. Is there an option to use Nexus without TypeScript?

--- a/src/runtime/app.ts
+++ b/src/runtime/app.ts
@@ -46,7 +46,7 @@ export interface App {
   use(plugin: Plugin.Plugin): void
   /**
    * Run this to gather the final state of all Nexus api interactions. This method
-   * is experimental. It provides experimental support for Nextjs integration.
+   * is experimental. It provides experimental support for NextJS integration.
    *
    * In a regular Nexus app, you should not need to use this method.
    *

--- a/src/runtime/server/server.ts
+++ b/src/runtime/server/server.ts
@@ -71,6 +71,10 @@ export function create(appState: AppState) {
       get graphql() {
         return (
           assembledGuard(appState, 'app.server.handlers.graphql', () => {
+            if (Boolean(settings.data.cors)) {
+              log.warn('CORS does not work for serverless handlers. Settings will be ignored.')
+            }
+
             return createRequestHandlerGraphQL(
               appState.assembled!.schema,
               appState.assembled!.createContext,
@@ -115,7 +119,11 @@ export function create(appState: AppState) {
             : false,
         })
 
-        state.apolloServer.applyMiddleware({ app: express, path: settings.data.path })
+        state.apolloServer.applyMiddleware({
+          app: express,
+          path: settings.data.path,
+          cors: settings.data.cors,
+        })
 
         return { createContext }
       },

--- a/src/runtime/server/settings.ts
+++ b/src/runtime/server/settings.ts
@@ -3,6 +3,7 @@ import { CorsOptions as OriginalCorsOption } from 'cors'
 import * as Process from '../../lib/process'
 import * as Utils from '../../lib/utils'
 import { log as serverLogger } from './logger'
+import { LiteralUnion } from 'type-fest'
 
 const log = serverLogger.child('settings')
 
@@ -10,36 +11,50 @@ export type PlaygroundSettings = {
   settings?: Omit<Partial<Exclude<PlaygroundRenderPageOptions['settings'], undefined>>, 'general.betaUpdates'>
 }
 
+export type HTTPMethods = LiteralUnion<'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | 'HEAD', string>
+
 export type CorsSettings = {
+  /**
+   * Enable or disable CORS.
+   * 
+   * @default true
+   */
+  enabled?: boolean
   /**
    * Configures the Access-Control-Allow-Origin CORS header. Possible values:
    *
    * Boolean - set origin to true to reflect the request origin, as defined by req.header('Origin'), or set it to false to disable CORS.
+   * 
    * String - set origin to a specific origin. For example if you set it to "http://example.com" only requests from "http://example.com" will be allowed.
+   * 
    * RegExp - set origin to a regular expression pattern which will be used to test the request origin. If it's a match, the request origin will be reflected. For example the pattern /example\.com$/ will reflect any request that is coming from an origin ending with "example.com".
+   * 
    * Array - set origin to an array of valid origins. Each origin can be a String or a RegExp. For example ["http://example1.com", /\.example2\.com$/] will accept any request from "http://example1.com" or from a subdomain of "example2.com".
+   * 
    * Function - set origin to a function implementing some custom logic. The function takes the request origin as the first parameter and a callback (called as callback(err, origin), where origin is a non-function value of the origin option) as the second.
    *
    */
-  origin?: OriginalCorsOption['origin']
+  origin?: OriginalCorsOption['origin'] // TODO: Improve function interface with promise-based callback
   /**
    * Configures the Access-Control-Allow-Methods CORS header.
-   *
-   * Expects a comma-delimited string (ex: 'GET,PUT,POST') or an array (ex: ['GET', 'PUT', 'POST']).
+   * 
+   * @example ['GET', 'PUT', 'POST']
    */
-  methods?: string | string[]
+  methods?: string | HTTPMethods[]
   /**
    * Configures the Access-Control-Allow-Headers CORS header.
    *
-   * Expects a comma-delimited string (ex: 'Content-Type,Authorization') or an array (ex: ['Content-Type', 'Authorization']).
    * If not specified, defaults to reflecting the headers specified in the request's Access-Control-Request-Headers header.
+   * 
+   * @example ['Content-Type', 'Authorization']
    */
   allowedHeaders?: string | string[]
   /**
    * Configures the Access-Control-Expose-Headers CORS header.
    *
-   * Expects a comma-delimited string (ex: 'Content-Range,X-Content-Range') or an array (ex: ['Content-Range', 'X-Content-Range']).
    * If not specified, no custom headers are exposed.
+   * 
+   * @example ['Content-Range', 'X-Content-Range']
    */
   exposedHeaders?: string | string[]
   /**
@@ -224,6 +239,18 @@ function validateGraphQLPath(path: string): string {
   return outputPath
 }
 
+export function corsSettings(newSettings: SettingsInput['cors']): SettingsData['cors'] {
+  if (typeof newSettings === 'boolean') {
+    return newSettings
+  }
+
+  if (!newSettings || newSettings.enabled === false) {
+    return false
+  }
+
+  return newSettings
+}
+
 /**
  * Mutate the settings data
  */
@@ -235,7 +262,7 @@ export function changeSettings(state: SettingsData, newSettings: SettingsInput):
   state.port = updatedSettings.port
   state.startMessage = updatedSettings.startMessage
   state.graphql = graphqlSettings(updatedSettings.graphql)
-  state.cors = updatedSettings.cors
+  state.cors = corsSettings(updatedSettings.cors)
 }
 
 export function createServerSettingsManager() {

--- a/src/runtime/server/settings.ts
+++ b/src/runtime/server/settings.ts
@@ -1,12 +1,67 @@
+import { PlaygroundRenderPageOptions } from 'apollo-server-express'
+import { CorsOptions as OriginalCorsOption } from 'cors'
 import * as Process from '../../lib/process'
 import * as Utils from '../../lib/utils'
-import { PlaygroundRenderPageOptions } from 'apollo-server-express'
 import { log as serverLogger } from './logger'
 
 const log = serverLogger.child('settings')
 
 export type PlaygroundSettings = {
   settings?: Omit<Partial<Exclude<PlaygroundRenderPageOptions['settings'], undefined>>, 'general.betaUpdates'>
+}
+
+export type CorsSettings = {
+  /**
+   * Configures the Access-Control-Allow-Origin CORS header. Possible values:
+   *
+   * Boolean - set origin to true to reflect the request origin, as defined by req.header('Origin'), or set it to false to disable CORS.
+   * String - set origin to a specific origin. For example if you set it to "http://example.com" only requests from "http://example.com" will be allowed.
+   * RegExp - set origin to a regular expression pattern which will be used to test the request origin. If it's a match, the request origin will be reflected. For example the pattern /example\.com$/ will reflect any request that is coming from an origin ending with "example.com".
+   * Array - set origin to an array of valid origins. Each origin can be a String or a RegExp. For example ["http://example1.com", /\.example2\.com$/] will accept any request from "http://example1.com" or from a subdomain of "example2.com".
+   * Function - set origin to a function implementing some custom logic. The function takes the request origin as the first parameter and a callback (called as callback(err, origin), where origin is a non-function value of the origin option) as the second.
+   *
+   */
+  origin?: OriginalCorsOption['origin']
+  /**
+   * Configures the Access-Control-Allow-Methods CORS header.
+   *
+   * Expects a comma-delimited string (ex: 'GET,PUT,POST') or an array (ex: ['GET', 'PUT', 'POST']).
+   */
+  methods?: string | string[]
+  /**
+   * Configures the Access-Control-Allow-Headers CORS header.
+   *
+   * Expects a comma-delimited string (ex: 'Content-Type,Authorization') or an array (ex: ['Content-Type', 'Authorization']).
+   * If not specified, defaults to reflecting the headers specified in the request's Access-Control-Request-Headers header.
+   */
+  allowedHeaders?: string | string[]
+  /**
+   * Configures the Access-Control-Expose-Headers CORS header.
+   *
+   * Expects a comma-delimited string (ex: 'Content-Range,X-Content-Range') or an array (ex: ['Content-Range', 'X-Content-Range']).
+   * If not specified, no custom headers are exposed.
+   */
+  exposedHeaders?: string | string[]
+  /**
+   * Configures the Access-Control-Allow-Credentials CORS header.
+   *
+   * Set to true to pass the header, otherwise it is omitted.
+   */
+  credentials?: boolean
+  /**
+   * Configures the Access-Control-Max-Age CORS header.
+   *
+   * Set to an integer to pass the header, otherwise it is omitted.
+   */
+  maxAge?: number
+  /**
+   * Pass the CORS preflight response to the next handler.
+   */
+  preflightContinue?: boolean
+  /**
+   * Provides a status code to use for successful OPTIONS requests, since some legacy browsers (IE11, various SmartTVs) choke on 204.
+   */
+  optionsSuccessStatus?: number
 }
 
 export type GraphqlSettings = {
@@ -33,15 +88,31 @@ export type SettingsInput = {
    * production, without some kind of security/access control, you will almost
    * certainly want this disabled.
    *
-   * To learn more about GraphQL Playgorund see
+   * To learn more about GraphQL Playground see
    * https://github.com/prisma-labs/graphql-playground
    */
   playground?: boolean | PlaygroundSettings
   /**
+   * Enable CORS for your server
+   *
+   * When true is passed, the default config is the following:
+   *
+   * ```
+   * {
+   *   "origin": "*",
+   *   "methods": "GET,HEAD,PUT,PATCH,POST,DELETE",
+   *   "preflightContinue": false,
+   *   "optionsSuccessStatus": 204
+   * }
+   * ```
+   *
+   * @default false
+   */
+  cors?: boolean | CorsSettings
+  /**
    * The path on which the GraphQL API should be served.
    *
-   * @default
-   * /graphql
+   * @default /graphql
    */
   path?: string
   /**
@@ -55,10 +126,14 @@ export type SettingsInput = {
   graphql?: GraphqlSettings
 }
 
-export type SettingsData = Omit<Utils.DeepRequired<SettingsInput>, 'host' | 'playground' | 'graphql'> & {
+export type SettingsData = Omit<
+  Utils.DeepRequired<SettingsInput>,
+  'host' | 'playground' | 'graphql' | 'cors'
+> & {
   host: string | undefined
   playground: false | Required<PlaygroundSettings>
   graphql: Required<GraphqlSettings>
+  cors: boolean | CorsSettings
 }
 
 export const defaultPlaygroundPath = '/graphql'
@@ -104,6 +179,7 @@ export const defaultSettings: () => Readonly<SettingsData> = () => {
     },
     playground: process.env.NODE_ENV === 'production' ? false : defaultPlaygroundSettings(),
     path: '/graphql',
+    cors: false,
     graphql: defaultGraphqlSettings(),
   }
 }
@@ -159,6 +235,7 @@ export function changeSettings(state: SettingsData, newSettings: SettingsInput):
   state.port = updatedSettings.port
   state.startMessage = updatedSettings.startMessage
   state.graphql = graphqlSettings(updatedSettings.graphql)
+  state.cors = updatedSettings.cors
 }
 
 export function createServerSettingsManager() {

--- a/website/content/040-api/01-nexus/04-settings.mdx
+++ b/website/content/040-api/01-nexus/04-settings.mdx
@@ -17,6 +17,7 @@ Use the settings to centrally configure various aspects of the various component
     host?: string
     path?: string
     playground?: boolean | PlaygroundSettings
+    cors?: boolean | CorsSettings
   }
   schema?: {
     nullable?: {
@@ -50,7 +51,7 @@ Change your app's current settings. This is an additive API meaning it can be ca
 
 #### `server.playground`
 
-```
+```ts
 boolean | PlaygroundSettings
 ```
 
@@ -62,7 +63,7 @@ _Default_
 
 #### `server.playground.settings`
 
-```
+```ts
 {
   'request.credentials': string;
   'tracing.hideTracingResponse': boolean;
@@ -79,7 +80,7 @@ Configure the Playground settings
 
 _Default_
 
-```
+```ts
 {
   'request.credentials': 'omit',
   'tracing.hideTracingResponse': true,
@@ -94,7 +95,7 @@ _Default_
 
 #### `server.port`
 
-```
+```ts
 number
 ```
 
@@ -109,7 +110,7 @@ _Default_
 
 #### `server.host`
 
-```
+```ts
 string
 ```
 
@@ -122,7 +123,7 @@ _Default_
 
 #### `server.path`
 
-```
+```ts
 string
 ```
 
@@ -132,9 +133,96 @@ _Default_
 
 `/graphql`
 
+#### `server.cors`
+
+```ts
+boolean | {
+  /**
+   * Enable or disable CORS.
+   *
+   * @default true
+   */
+  enabled?: boolean
+  /**
+   * Configures the Access-Control-Allow-Origin CORS header. Possible values:
+   *
+   * Boolean - set origin to true to reflect the request origin, as defined by req.header('Origin'), or set it to false to disable CORS.
+   *
+   * String - set origin to a specific origin. For example if you set it to "http://example.com" only requests from "http://example.com" will be allowed.
+   *
+   * RegExp - set origin to a regular expression pattern which will be used to test the request origin. If it's a match, the request origin will be reflected. For example the pattern /example\.com$/ will reflect any request that is coming from an origin ending with "example.com".
+   *
+   * Array - set origin to an array of valid origins. Each origin can be a String or a RegExp. For example ["http://example1.com", /\.example2\.com$/] will accept any request from "http://example1.com" or from a subdomain of "example2.com".
+   *
+   * Function - set origin to a function implementing some custom logic. The function takes the request origin as the first parameter and a callback (called as callback(err, origin), where origin is a non-function value of the origin option) as the second.
+   *
+   */
+  origin?: OriginalCorsOption['origin'] // TODO: Improve function interface with promise-based callback
+  /**
+   * Configures the Access-Control-Allow-Methods CORS header.
+   *
+   * @example ['GET', 'PUT', 'POST']
+   */
+  methods?: string | HTTPMethods[]
+  /**
+   * Configures the Access-Control-Allow-Headers CORS header.
+   *
+   * If not specified, defaults to reflecting the headers specified in the request's Access-Control-Request-Headers header.
+   *
+   * @example ['Content-Type', 'Authorization']
+   */
+  allowedHeaders?: string | string[]
+  /**
+   * Configures the Access-Control-Expose-Headers CORS header.
+   *
+   * If not specified, no custom headers are exposed.
+   *
+   * @example ['Content-Range', 'X-Content-Range']
+   */
+  exposedHeaders?: string | string[]
+  /**
+   * Configures the Access-Control-Allow-Credentials CORS header.
+   *
+   * Set to true to pass the header, otherwise it is omitted.
+   */
+  credentials?: boolean
+  /**
+   * Configures the Access-Control-Max-Age CORS header.
+   *
+   * Set to an integer to pass the header, otherwise it is omitted.
+   */
+  maxAge?: number
+  /**
+   * Pass the CORS preflight response to the next handler.
+   */
+  preflightContinue?: boolean
+  /**
+   * Provides a status code to use for successful OPTIONS requests, since some legacy browsers (IE11, various SmartTVs) choke on 204.
+   */
+  optionsSuccessStatus?: number
+}
+```
+
+Enable CORS for your server.
+
+When true is passed, the default config is the following:
+
+```ts
+{
+  "origin": "*",
+  "methods": "GET,HEAD,PUT,PATCH,POST,DELETE",
+  "preflightContinue": false,
+  "optionsSuccessStatus": 204
+}
+```
+
+_Default_
+
+false
+
 #### `schema.nullable.inputs`
 
-```
+```ts
 boolean
 ```
 
@@ -146,7 +234,7 @@ _Default_
 
 #### `schema.nullable.outputs`
 
-```
+```ts
 boolean
 ```
 
@@ -158,7 +246,7 @@ _Default_
 
 #### `schema.generateGraphQLSDLFile`
 
-```
+```ts
 false | string
 ```
 
@@ -172,7 +260,7 @@ _Default_
 
 #### `schema.rootTypingsGlobPattern`
 
-```
+```ts
 string
 ```
 

